### PR TITLE
OCPBUGS-16128:  daemon: Copy matching binary to host, re-exec with it

### PR DESCRIPTION
**Pull request by cgwalters**
4.13: use rhel9 builder for daemon binary

While attempting to switch builder to rhel9 base, the nmstate binary from
rhel9 fails. With that change reverted for now, let's try just building
the daemon binary on rhel9, so when it gets copied to host for
firstboot, it matches the on-node os version (rhel9)

For https://issues.redhat.com/browse/TRT-1143

(walters: I also snuck in a cleanup to use `chcon --reference` here
 because it avoids hardcoding a SELinux context)

---

build-sys: Drop setting CGO_ENABLED=0

We must dynamically link in general.

---

daemon: Run firstboot as a container image too

As the TODOs say, I am now no longer sure why we did the
"copy self to host" instead of always running from the container.

IOW, this makes firstboot work *exactly the same* as the daemonset
running after kube is up.

We really need to do this now because now we need to ensure
that the code flows here in a dynamic linking world properly
handle matching the target host system.

---

daemon: Copy matching binary to host, re-exec with it

In the general case as we want to dynamically link our Go code,
we simply must have a binary matching the host system in general.

We're already building both into the image; this changes our
entrypoint logic to always do:

- copy matching binary to /rootfs/run/bin/machine-config-daemon
- chroot /rootfs
- execve(/run/bin/machine-config-daemon, --root-mount=/ ...)

The presence of both `--root-mount=/` and a "guard environment variable"
avoid recursion.

---

### CODE REVIEW

**Comment by cgwalters** [comment]
OK, fixed some bugs, still fixing more

File: pkg/daemon/osrelease/osrelease.go
Comment (cgwalters) [inline pkg/daemon/osrelease/osrelease.go:87]: To my great surprise `IsEL()` actually returns false for non-coreos rhel (i.e. the base container image).

File: Dockerfile
Comment (cgwalters) [inline Dockerfile:11]: cc OKD folks - is this a problem?  Do we need distinct dockerfiles?

Comment (LorbusChris) [reply Dockerfile:11]: This should be fine for Prow builds (i.e. the official OKD builds). It might cause problems when using https://github.com/okd-project/okd-payload-pipeline/ to builds outside of Prow, but we should be able to work around it there, too.

cc @openshift/okd-maintainers @aleskandro

**Comment by sinnykumari** [comment]
These tests are passing on local machine, rerunning them
/test verify
/test unit

**Comment by sinnykumari** [comment]
PR as it is looks good to go for initial round of testing. Removing draft  to get all necessary tests running.

**Comment by sinnykumari** [comment]
putting hold as I will be running some tests first, also needes qe-approved label
/hold

**Comment by sinnykumari** [comment]
Failure doesn't seem to be related to this PR as I see similar error in other recent run, also must-gather doesn't have any useful log. Checking with other team to see if it is a known issue.

**Comment by sinnykumari** [comment]
ok, found some relevant log in [install log ](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3799/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1681220984276783104/artifacts/e2e-gcp-op/ipi-install-install/artifacts/) , looking at the log, relevant error is :
```
I0718 08:57:55.560546       1 daemon.go:456] container is rhel8, target is rhel9
I0718 08:57:55.594419       1 daemon.go:523] Invoking re-exec /run/bin/machine-config-daemon
I0718 08:57:55.636963       1 update.go:1893] Running: systemctl daemon-reload
W0718 08:57:55.646723       1 firstboot_complete_machineconfig.go:65] error: error initializing rpm-ostree: error running systemctl daemon-reload: Failed to connect to bus: No data available
: exit status 1
I0718 08:57:55.646904       1 firstboot_complete_machineconfig.go:66] Sleeping 1 minute for retry
W0718 08:58:55.763512       1 firstboot_complete_machineconfig.go:65] error: error reading osImageURL from rpm-ostree: exit status 1
```

**Comment by cgwalters** [comment]
Ah right looks like we need `--pid=host` too.  In general, we want to make running via `podman` be the *exact same configuration* as our daemonset, which has this.  See https://github.com/openshift/machine-config-operator/blob/d0c0670f38dcf95d1d267879614ad09e37e58d51/manifests/machineconfigdaemon/daemonset.yaml#L84

**Comment by sinnykumari** [comment]
All tests are passing \o/
I also ran some local testing including scale-up of a node from 4.12 based bootimage and it went well.

This PR is part of MCO side changes to fix OCPBUGS-16128. For complete fix of the bug after this PR merged, we will need Revert of https://github.com/openshift-eng/ocp-build-data/pull/3186 to use fips enabled golang

Also running payload tests to make sure this PR doesn't break anything
/payload 4.14 nightly blocking

@sergiordlr would you like to run some tests and see we are not regressing?

**Comment by sinnykumari** [comment]
Most of the periodice jobs has passed, rerunning failed one
/payload-job periodic-ci-openshift-release-master-ci-4.14-e2e-azure-ovn-upgrade periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ipi-sdn-bm

**Comment by sinnykumari** [comment]
This time both tests failed due to "Error: Unable to find a match: nmstate-2.2.9-6.rhaos4.14.el8" , this was not the case in previous run. Re-running
/payload-job periodic-ci-openshift-release-master-ci-4.14-e2e-azure-ovn-upgrade periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ipi-sdn-bm

**Comment by sinnykumari** [comment]
we haven't made any changes in this PR related to nmstate version pinning, so failing payload test shouldn't be  related to this PR.
Anyway, re-running a passing test to test my theory
/payload-job periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ipi-ovn-ipv6

**Comment by sinnykumari** [comment]
yep, periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ipi-ovn-ipv6 now failed too. Looks like nmstate-2.2.9-6.rhaos4.14.el8 may have got untagged recently or somethingelse?

**Comment by sergiordlr** [comment]
Verification steps:

- **Installed using IPI on AWS (template: private-templates/functionality-testing/aos-4_14/ipi-on-aws/versioned-installer)**

We can see that the new logic is executed (container is rhel8, target is rhel9):
```
]$  oc debug -q  node/$(oc get nodes -l node-role.kubernetes.io/master -ojsonpath="{.items[0].metadata.name}") -- chroot /host journalctl -u machine-config-daemon-firstboot.service | head
Jul 19 12:17:55 ip-10-0-149-9 systemd[1]: Starting Machine Config Daemon Firstboot...
Jul 19 12:17:55 ip-10-0-149-9 sh[2044]: sed: can't read /etc/yum.repos.d/*.repo: No such file or directory
Jul 19 12:17:55 ip-10-0-149-9 podman[2045]: W0719 12:17:55.547905       1 daemon.go:1626] Failed to persist NIC names: open /rootfs/etc/systemd/network: no such file or directory
Jul 19 12:17:55 ip-10-0-149-9 podman[2114]: I0719 12:17:55.752387    2149 daemon.go:456] container is rhel8, target is rhel9
Jul 19 12:17:55 ip-10-0-149-9 podman[2114]: I0719 12:17:55.775556    2149 daemon.go:523] Invoking re-exec /run/bin/machine-config-daemon
Jul 19 12:17:55 ip-10-0-149-9 podman[2114]: I0719 12:17:55.802400    2149 update.go:1893] Running: systemctl daemon-reload
Jul 19 12:17:55 ip-10-0-149-9 podman[2114]: I0719 12:17:55.957735    2149 rpm-ostree.go:85] Enabled workaround for bug 2111817
Jul 19 12:17:56 ip-10-0-149-9 podman[2114]: I0719 12:17:56.076567    2149 daemon.go:261] Booted osImageURL:  (414.92.202307070025-0) ea334ea715125202a4adeeed53d572c0d6c11703fb931e3e53b3afba3d01d350
Jul 19 12:17:56 ip-10-0-149-9 podman[2114]: I0719 12:17:56.076853    2149 rpm-ostree.go:382] Running captured: rpm-ostree --version
Jul 19 12:17:56 ip-10-0-149-9 podman[2114]: I0719 12:17:56.089989    2149 daemon.go:1087] rpm-ostree has container feature
```


No failures found in the installation


We executed these test cases and they passed:
"[sig-mco] MCO scale Author:sregidor-NonHyperShiftHOST-NonPreRelease-LongDuration-High-63894-Scaleup using 4.1 cloud image[Disruptive] [Serial]"
"[sig-mco] MCO scale Author:sregidor-NonPreRelease-High-52822-Create new config resources with 2.2.0 ignition boot image nodes [Disruptive] [Serial]"
"[sig-mco] MCO Layering Author:sregidor-ConnectedOnly-VMonly-Longduration-NonPreRelease-Critical-54085-Update osImage changing /etc /usr and rpm [Disruptive] [Serial]"

When we use the 4.1 old cloud image in test TC-63894 we can see this log: `not chrooting for source=rhel-%s target=rhel-%s84` in ther machine-config-daemon-firstboot service:

```
$ oc debug -q node/ip-10-0-156-58.us-east-2.compute.internal -- chroot /host journalctl -u machine-config-daemon-firstboot.service | head
Jul 19 13:55:43 ip-10-0-156-58 systemd[1]: Starting Machine Config Daemon Firstboot...
Jul 19 13:55:43 ip-10-0-156-58 sh[1364]: sed: can't read /etc/yum.repos.d/*.repo: No such file or directory
Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.355256    1471 daemon.go:471] not chrooting for source=rhel-%s target=rhel-%s84
Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.375947    1471 daemon.go:523] Invoking re-exec /run/bin/machine-config-daemon
Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.397829    1471 update.go:1893] Running: systemctl daemon-reload
Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.478504    1471 rpm-ostree.go:85] Enabled workaround for bug 2111817
Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.500709    1471 daemon.go:261] Booted osImageURL: docker-registry-default.cloud.registry.upshift.redhat.com/redhat-coreos/ootpa@sha256:683a6a866a8ec789fedb5da63b6a2ff68c1b0788ec90e7def778f0c4c13197a4 (410.8.20190520.0) 0e5effbb2219acc8f4b15f9
```


The only problem that we see when adding a 4.1 old image is that the log message `not chrooting for source=rhel-%s target=rhel-%s84` seems to be malformed.


- **Installed using IPI on GCP. (template: private-templates/functionality-testing/aos-4_14/ipi-on-gcp/versioned-installer)**

No problems found


- **Installed using UPI on AWS with fips. (template:  private-templates/functionality-testing/aos-4_14/upi-on-aws/versioned-installer-fips-ci)**

Added one yum based RHEL8 node to the cluster. This is the first boot output for this node:
```
$ oc debug -q  node/$(oc get nodes -l node.openshift.io/os_id=rhel -ojsonpath="{.items[0].metadata.name}") -- chroot /host journalctl -u machine-config-daemon-firstboot.service 
-- Logs begin at Wed 2023-07-19 13:30:01 UTC, end at Wed 2023-07-19 13:38:17 UTC. --
-- No entries --

```

In the MCD logs we cann see (not chrooting for source=rhel-%s target=rhel-%s88):

```
I0719 13:31:05.852986    5540 start.go:61] Version: machine-config-daemon-4.6.0-202006240615.p0-2178-g8f10d68b-dirty (8f10d68bb2b64b9a92c07f9a8d627816f085b377)
I0719 13:31:05.853277    5540 update.go:1893] Running: mount --rbind /run/secrets /rootfs/run/secrets
I0719 13:31:06.055161    5540 update.go:1893] Running: mount --rbind /usr/bin /rootfs/run/machine-config-daemon-bin
I0719 13:31:06.067260    5540 daemon.go:471] not chrooting for source=rhel-%s target=rhel-%s88
I0719 13:31:06.125431    5540 daemon.go:523] Invoking re-exec /run/bin/machine-config-daemon
I0719 13:31:06.283137    5540 start.go:61] Version: machine-config-daemon-4.6.0-202006240615.p0-2178-g8f10d68b-dirty (8f10d68bb2b64b9a92c07f9a8d627816f085b377)
I0719 13:31:06.283286    5540 update.go:1893] Running: mount --rbind /run/secrets /run/secrets
I0719 13:31:06.287142    5540 update.go:1893] Running: mount --rbind /usr/bin /run/machine-config-daemon-bin
I0719 13:31:06.291074    5540 start.go:124] overriding kubernetes api to https://api-int.sregidor-upiawsvgf.qe.devcluster.openshift.com:6443
I0719 13:31:06.294153    5540 metrics.go:74] Registering Prometheus metrics
I0719 13:31:06.294252    5540 metrics.go:81] Starting metrics listener on 127.0.0.1:8797
W0719 13:31:06.329782    5540 controller_context.go:111] unable to get owner reference (falling back to namespace): pods is forbidden: User "system:serviceaccount:openshift-machine-config-operator:machine-config-daemon" cannot list resource "pods" in API group "" in the namespace "machine-config-daemon"
I0719 13:31:06.331865    5540 simple_featuregate_reader.go:171] Starting feature-gate-detector
I0719 13:31:06.332813    5540 writer.go:93] NodeWriter initialized with credentials from /var/lib/kubelet/kubeconfig
```

No problems found


- **Installed using UPI on baremetal ( private-templates/functionality-testing/aos-4_14/upi-on-baremetal/versioned-installer-packet with fips_enable: true)**

No problems found


- **Upgrade from 4.12 (template: private-templates/functionality-testing/aos-4_12/ipi-on-aws/versioned-installer)**

This is the upgrade history:

```
    history:
    - acceptedRisks: |-
        Target release version="" image="registry.build05.ci.openshift.org/ci-ln-khh77vk/release:latest" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified
        Forced through blocking failures: Multiple precondition checks failed:
        * Precondition "ClusterVersionUpgradeable" failed because of "PoolUpdating": Cluster operator machine-config should not be upgraded between minor versions: One or more machine config pools are updating, please see `oc get mcp` for further details
        * Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.0-0.nightly-2023-07-18-211451 to 4.14.0-0.ci.test-2023-07-19-100409-ci-ln-khh77vk-latest is unknown.
      completionTime: "2023-07-19T16:33:45Z"
      image: registry.build05.ci.openshift.org/ci-ln-khh77vk/release:latest
      startedTime: "2023-07-19T15:25:36Z"
      state: Completed
      verified: false
      version: 4.14.0-0.ci.test-2023-07-19-100409-ci-ln-khh77vk-latest
    - acceptedRisks: |-
        Target release version="" image="registry.ci.openshift.org/ocp/release:4.13.0-0.nightly-2023-07-18-211451" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified
        Forced through blocking failures: Multiple precondition checks failed:
        * Precondition "ClusterVersionUpgradeable" failed because of "AdminAckRequired": Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions.
        * Precondition "EtcdRecentBackup" failed because of "ControllerStarted": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required
        * Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.12.0-0.nightly-2023-07-19-020444 to 4.13.0-0.nightly-2023-07-18-211451 is unknown.
      completionTime: "2023-07-19T15:19:56Z"
      image: registry.ci.openshift.org/ocp/release:4.13.0-0.nightly-2023-07-18-211451
      startedTime: "2023-07-19T14:13:23Z"
      state: Completed
      verified: false
      version: 4.13.0-0.nightly-2023-07-18-211451
    - completionTime: "2023-07-19T13:54:11Z"
      image: registry.ci.openshift.org/ocp/release@sha256:67fdd720ff787df8d5225f59ef4212aca414dfee9852fb50cc731b86aa0ee2c4
      startedTime: "2023-07-19T13:37:08Z"
      state: Completed
      verified: false
      version: 4.12.0-0.nightly-2023-07-19-020444
```

We scaled up a worker node after the upgrade

No problem found.



- **Summary:**

Only two things to notice:

1. The log message when exuting souce=8 and target!=9 is malformed: 

`Jul 19 13:55:50 ip-10-0-156-58 podman[1435]: I0719 13:55:50.355256    1471 daemon.go:471] not chrooting for source=rhel-%s target=rhel-%s84`


2. The firstboot service exists with an error message, instead of a controlled message.

Before this PR, the first boot service exited like this:

```
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal machine-config-daemon[2643]: I0719 09:01:45.565770    2643 update.go:1908] reboot successful
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal root[3220]: machine-config-daemon[2643]: reboot successful
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal systemd[1]: Finished Machine Config Daemon Firstboot.
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal systemd[1]: machine-config-daemon-firstboot.service: Deactivated successfully.
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal systemd[1]: Stopped Machine Config Daemon Firstboot.
Jul 19 09:01:45 sregidor-v4-nfktt-master-0.c.openshift-qe.internal systemd[1]: machine-config-daemon-firstboot.service: Consumed 5.952s CPU time.
-- Boot fa25f3c060584d37a7bcecddd9d193d9 --

```

Now it exits reporting a failure due to termination (machine-config-daemon-firstboot.service: Failed with result 'signal'.):

```
Jul 19 12:18:33 ip-10-0-149-9 podman[2114]: I0719 12:18:33.627466    2149 update.go:1908] Rebooting node
Jul 19 12:18:33 ip-10-0-149-9 podman[2114]: I0719 12:18:33.629012    2149 update.go:1938] Removing SIGTERM protection
Jul 19 12:18:33 ip-10-0-149-9 podman[2114]: I0719 12:18:33.629061    2149 update.go:1908] initiating reboot: Completing firstboot provisioning to rendered-master-2723ad6eba300243c6393a1981eb7b21
Jul 19 12:18:33 ip-10-0-149-9 podman[2114]: I0719 12:18:33.647857    2149 update.go:1908] reboot successful
Jul 19 12:18:33 ip-10-0-149-9 systemd[1]: machine-config-daemon-firstboot.service: Main process exited, code=killed, status=15/TERM
Jul 19 12:18:33 ip-10-0-149-9 systemd[1]: machine-config-daemon-firstboot.service: Failed with result 'signal'.
Jul 19 12:18:33 ip-10-0-149-9 systemd[1]: Stopped Machine Config Daemon Firstboot.
-- Boot 48bdb23c02c24a0a8020514f4a5e3f2f --

```

Both issues look like cosmetic. Please, could you confirm that?

The malformed log is clearly a cosmetic thing, but is the new way to finish the firstboot service only a cosmetic thing as well? is it any problem?

**Comment by sinnykumari** [comment]
Thanks Sergio for verifying!
1. Yep, it is a cosmetic change(to not run Un-necssary all ci tests) and I am intending to fix it in follow-up PR as this is just formatting issue but we have all data in the log message. 
2. I think this one could be as a result of us moving mcd firstbootcomplete service moving inside container and reboot initiated from there. This shouldn't be an issue. @mrunalp @cgwalters can correct me.

File: pkg/daemon/daemon.go
Comment (cdoern) [inline pkg/daemon/daemon.go:524]: I could be wrong, but doesn't syscall.Exec return a kind of useless error unless you parse it further?

File: pkg/daemon/daemon.go
Comment (cdoern) [inline pkg/daemon/daemon.go:454]: might be contrived but this just popped into my head
```
if sourceMajor != "" {
  sourceBinarySuffix = ".rhel"+sourceMajor
  klog.Info(fmt.Sprintf("targeting rhel%s which is a different version than the container rhel%s",  targetMajor, sourceMajor))
} else {
  klog.Info("not chrooting for source=rhel-%s target=rhel-%s", sourceMajor, targetMajor)
}
```

File: cmd/machine-config-daemon/start.go
Comment (cdoern) [inline cmd/machine-config-daemon/start.go:54]: might be too much of a functionality change but could this be a `runE` rather than a `run` with cobra so we can return an error rather than klog.Fatal ?

**Review by cdoern — APPROVED** [review]
Looks good, just adding my two cents

**Comment by cgwalters** [comment]
> I think this one could be as a result of us moving mcd firstbootcomplete service moving inside container and reboot initiated from there. This shouldn't be an issue. @mrunalp @cgwalters can correct me.

Right.  What we could do is spawn off `systemd-run reboot` instead, which is a pattern we've used in other places.  But that's still racy.

Ultimately, the fix for this is quite complex - it's basically:

- `flock -x /run/reboot-lock`
- `systemd-run /bin/sh -c 'flock -x /run/reboot-lock reboot`
- `exit 0` (releasing the lock)

This isn't *perfect* because there's still a race between when the process exits and pid1 gets the notification, but I think systemd/pid1 will handle this by observing that the process has already exited after it sends SIGTERM.

**Comment by cgwalters** [comment]
Oh huh...looks like in newer systemd there is a race-free `SuccessAction=reboot`.

**Comment by sinnykumari** [comment]
since this is not impacting any functionality, perhaps we can improve on this in a follow-up PR if needed

**Comment by sinnykumari** [comment]
> Oh huh...looks like in newer systemd there is a race-free `SuccessAction=reboot`.

hmm, but but this may fail for older bootimages due to older systemd version?

File: pkg/daemon/daemon.go
Comment (sinnykumari) [reply pkg/daemon/daemon.go:524]: umm, in our case we don't need to further parse the error. we are simply catching the error and printing it in caller.

File: pkg/daemon/daemon.go
Comment (sinnykumari) [reply pkg/daemon/daemon.go:454]: baseversion comes from /etc/os-release which is part of standard package. We can rely that it won't break otherwise we would be breaking multiple consumers of os-release file.

**Comment by sinnykumari** [comment]
@cgwalters do you think we should implement in this PR what is suggested in https://github.com/openshift/machine-config-operator/pull/3799#issuecomment-1642614283 or https://github.com/openshift/machine-config-operator/pull/3799#issuecomment-1642626235 ?

**Comment by cgwalters** [comment]
IMO no, it's just cosmetic.

**Comment by sergiordlr** [comment]
It is not necessary now, but since we pre-merge verified it, we add the qe-approved label anyway.

/label qe-approved

**Pull request merged**
